src/random_ray/iteration.cpp

```diff
@@ -0,0 +1,660 @@
+    parallel_fill<float>(random_ray::scalar_flux_new, 0.0f);
+    parallel_fill<double>(random_ray::volume, 0.0);
+    parallel_fill<int>(random_ray::was_hit, 0);
+
+    // Start timer for transport
+    simulation::time_transport.start();
+
+// Transport sweep over all random rays for the iteration
+#pragma omp parallel for schedule(runtime)                                     \
+  reduction(+ : total_geometric_intersections)
+    for (int i = 0; i < simulation::work_per_rank; i++) {
+      RandomRay ray;
+      ray.initialize_ray(i, sampling_source);
+      total_geometric_intersections += ray.transport_history_based_single_ray();
+    }
+
+    simulation::time_transport.stop();
+
+    // If using multiple MPI ranks, perform all reduce on all transport results
+    all_reduce_random_ray_batch_results(mapped_all_tallies);
+
+    // Normalize scalar flux and update volumes
+    normalize_scalar_flux_and_volumes();
+
+    // Add source to scalar flux, compute number of FSR hits
+    int64_t n_hits = add_source_to_scalar_flux();
+
+    // Compute random ray k-eff
+    k_eff = compute_k_eff(k_eff);
+
+    // Store random ray k-eff into OpenMC's native k-eff variable
+    global_tally_tracklength = k_eff;
+
+    // Execute all tallying tasks, if this is an active batch
+    if (simulation::current_batch > settings::n_inactive && mpi::master) {
+
+      // Generate mapping between source regions and tallies
+      if (!mapped_all_tallies) {
+        mapped_all_tallies = convert_source_regions_to_tallies();
+      }
+
+      // Use above mapping to contribute FSR flux data to appropriate tallies
+      random_ray_tally();
+
+// Add this iteration's scalar flux estimate to final accumulated estimate
+#pragma omp parallel for
+      for (int64_t se = 0; se < random_ray::n_source_elements; se++) {
+        random_ray::scalar_flux_final[se] += random_ray::scalar_flux_new[se];
+      }
+    }
+
+    // Set phi_old = phi_new
+    random_ray::scalar_flux_old.swap(random_ray::scalar_flux_new);
+
+    // Check for any obvious insabilities/nans/infs
+    instability_check(n_hits, k_eff, avg_miss_rate);
+
+    // Finalize the current batch
+    finalize_generation();
+    finalize_batch();
+  } // End random ray power iteration loop
+
+  openmc::simulation::time_total.stop();
+
+  // Finalize OpenMC
+  openmc_simulation_finalize();
+
+  // Reduce number of intersections
+#ifdef OPENMC_MPI
+  if (mpi::n_procs > 1) {
+    uint64_t total_geometric_intersections_reduced = 0;
+    MPI_Reduce(&total_geometric_intersections,
+      &total_geometric_intersections_reduced, 1, MPI_UNSIGNED_LONG, MPI_SUM, 0,
+      mpi::intracomm);
+    total_geometric_intersections = total_geometric_intersections_reduced;
+  }
+#endif
+
+  // Print random ray results
+  if (mpi::master) {
+    print_results_random_ray(
+      total_geometric_intersections, avg_miss_rate / settings::n_batches);
+    if (model::plots.size() > 0) {
+      plot_3D_vtk();
+    }
+  }
+
+  return 0;
```

Comment:
Looks like this function should just be void return type. No scenario in which an error code is returned, right? Besides, that's not generally how error handling is done in OpenMC these days.